frame/staking/src/offchain_election.rs

```diff
@@ -155,19 +163,163 @@ pub fn get_balancing_iters<T: Trait>() -> usize {
 	}
 }
 
+/// Find the maximum `len` that a compact can have in order to fit into the block weight.
+///
+/// We start by assuming `voters == size.nominators`, i.e. all nominators are active in the
+/// solution. Then we do a binary search until we find the ideal value.
+///
+/// Thus, this only returns a value between zero and `size.nominators`.
+pub fn maximum_compact_len<W: crate::WeightInfo>(
+	winners_len: u32,
+	size: ElectionSize,
+	max_weight: Weight,
+) -> u32 {
+	use sp_std::cmp::Ordering;
+	if size.nominators < 1 {
+		return size.nominators;
+	}
+
+	let weight_with = |voters: u32| -> Weight {
+		W::submit_solution_better(
+			size.validators.into(),
+			size.nominators.into(),
+			voters,
+			winners_len,
+		)
+	};
+
+	let mut voters = size.nominators.max(2);
+	let mut step = voters;
+	loop {
```

Comment:
This loop will turn `log(size.nominators)` times anyway if `weight_with(size.nominators) < max_weight`, we should bound `voters` earlier:
```suggestion
	while voters <= size.nominators {
```
Although I'm not sure what should happen if `size.nominators == 1`, cause with that condition we won't enter the loop at all. Perhaps we need this `.max(2)` here as well.

Reply:
I added the condition elsewhere when we do `voters += step;`, I find that more expressive.

Reply:
I agree.

And Indeed we need an exception only for size.nominators == 1, effectively we must treat it as 2.